pkg/syncer/storage/storage_pvc_process.go

```diff
@@ -0,0 +1,240 @@
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to get downstream PersistentVolumeClaim: %w", err)
+	}
+
+	downstreamPVC, ok := downstreamPVCObject.(*corev1.PersistentVolumeClaim)
+	if !ok {
+		return fmt.Errorf("failed to assert object to a PersistentVolumeClaim: %T", downstreamPVCObject)
+	}
+
+	logger.V(1).Info("processing downstream PersistentVolumeClaim")
+
+	switch downstreamPVC.Status.Phase {
+	case corev1.ClaimBound:
+		// var updateObject bool
+
+		// Fetch the namespace locator annotation from the downstream PVC, this will be helpful to
+		// get the upstream PVC
+		nsLocatorJSON, ok := downstreamPVC.GetAnnotations()[shared.NamespaceLocatorAnnotation]
+		if !ok || nsLocatorJSON == "" {
+			return fmt.Errorf("downstream PersistentVolumeClaim %q does not have the %q annotation", downstreamPVC.Name, shared.NamespaceLocatorAnnotation)
+		}
+
+		// Convert the string from the annotation into a NamespaceLocator object
+		nsLocator := shared.NamespaceLocator{}
+		if err := json.Unmarshal([]byte(nsLocatorJSON), &nsLocator); err != nil {
+			return fmt.Errorf("failed to unmarshal namespace locator from downstream PersistentVolumeClaim %q: %w", downstreamPVC.Name, err)
+		}
+
+		// Get the PV corresponding to the PVC
+		downstreamPersistentVolumeName := downstreamPVC.Spec.VolumeName
+		downstreamPVObject, err := c.getDownstreamPersistentVolume(downstreamPersistentVolumeName)
+		if err != nil {
+			return fmt.Errorf("failed to get downstream PersistentVolume: %w", err)
+		}
+
+		downstreamPV, ok := downstreamPVObject.(*corev1.PersistentVolume)
+		if !ok {
+			return fmt.Errorf("failed to assert object to a PersistentVolume: %T", downstreamPVObject)
+		}
+
+		downstreamPV, err = addNsAllocatorToDownstreamPV(nsLocatorJSON, downstreamPV)
+		if err != nil {
+			return fmt.Errorf("failed to add namespace locator to downstream PersistentVolume: %w", err)
+		}
+
+		// Get upstream PVC from downstream PVC
+		upstreamPVCObject, err := c.getUpstreamPersistentVolumeClaim(nsLocator.Workspace, downstreamPVC.Name, nsLocator.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get upstream PersistentVolumeClaim: %w", err)
+		}
+
+		upstreamPVC, ok := upstreamPVCObject.(*corev1.PersistentVolumeClaim)
+		if !ok {
+			return fmt.Errorf("failed to assert object to a PersistentVolumeClaim: %T", upstreamPVCObject)
+		}
+
+		// Create a Json patch for the PV resource whose content would update the PVC reference to match the upstream PVC reference (namespace, uid, resource version, etc ...)
+		upstreamClaimRef := &corev1.ObjectReference{
+			Kind:            upstreamPVC.Kind,
+			Namespace:       upstreamPVC.Namespace,
+			Name:            upstreamPVC.Name,
+			UID:             upstreamPVC.UID,
+			APIVersion:      upstreamPVC.APIVersion,
+			ResourceVersion: upstreamPVC.ResourceVersion,
+		}
+
+		// If the object being reconciled changed as a result, update it.
```

Comment:
We have a patcher utility, let's use that

Reply:
Can you point me to it?

Reply:
Sure thing! Here: https://github.com/kcp-dev/kcp/blob/main/pkg/reconciler/committer/committer.go

Reply:
Confirmed offline that we need a committer that is not cluster-aware. So we can revisit this a followup PR.

Reply:
Actually, I went ahead and implemented something. I'll split into individual commits at the end of the review.